src/operator.rs

```diff
@@ -1,15 +1,43 @@
 use crate::array::Array;
 use crate::error::Error;
+use crate::hardware::Hardware;
 use crate::node::Node;
 use crate::result::Result;
 use crate::shape::Shape;
+use std::cell::RefCell;
+use std::ptr;
 
 /// Operator represents an individual computation process in the computation graph.
 pub(crate) trait Operator<'hw> {
     fn name(&self) -> String;
     fn input_size(&self) -> usize;
     fn perform_shape(&self, inputs: &[&Shape]) -> Result<Shape>;
+
+    fn perform_hardware(
+        &self,
+        inputs: &[&'hw RefCell<dyn Hardware>],
+    ) -> Result<&'hw RefCell<dyn Hardware>> {
+        // Most operations assumes that all inputs are on the same hardware.
```

Comment:
It seems this comment and the error messages in the code should be written as doc comments (`///`) because that sounds like requirements (or design decisions) callers need to follow.

```suggestion
        // Most operations assume that all inputs are on the same hardware.
```